internal/pg/postgres/password/scram.go

```diff
@@ -0,0 +1,259 @@
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+import (
+	"crypto/hmac"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"hash"
+	"unicode"
+	"unicode/utf8"
+
+	"github.com/xdg/stringprep"
+	"golang.org/x/crypto/pbkdf2"
+)
+
+// scramVerifierFormat is the format of the value that is stored by PostgreSQL
+// and follows the format:
+//
+// <DIGEST>$<ITERATIONS>:<SALT>$<STORED_KEY>:<SERVER_KEY>
+//
+// where:
+// DIGEST = SCRAM-SHA-256 (only value for now in PostgreSQL)
+// ITERATIONS = the number of iteratiosn to use for PBKDF2
+// SALT = the salt used as part of the PBKDF2, stored in base64
+// STORED_KEY = the hash of the client key, stored in base64
+// SERVER_KEY = the hash of the server key
+const scramVerifierFormat = "SCRAM-SHA-256$%d:%s$%s:%s"
+
+// These constants are defined as part of the PostgreSQL implementation for
+// SCRAM, but can be overriden by the user
+// https://git.postgresql.org/gitweb/?p=postgresql.git;a=blob;f=src/include/common/scram-common.h
+const (
+	// scramDefaultIterations is the number of iterations to make as part of the PBKDF2
+	// algorithm
+	scramDefaultIterations = 4096
+	// scramDefaultSaltLength is the length of the generated salt used in creating the
+	// hashed password
+	scramDefaultSaltLength = 16
+)
+
+// scramDefaultHash is the hashing algorithm to use
+var scramDefaultHash = sha256.New
+
+// the following are used as part of the SCRAM verifier generation
+var (
+	scramClientKeyMessage = []byte("Client Key")
+	scramServerKeyMessage = []byte("Server Key")
+)
+
+var (
+	// ErrSCRAMPasswordInvalid is returned when the password attributes are invalid
+	ErrSCRAMPasswordInvalid = errors.New(`invalid password attributes. must provide "password"`)
+	// ErrSCRAMSaltLengthInvalid is returned when the salt length is less than 1
+	ErrSCRAMSaltLengthInvalid = errors.New(`salt length must be at least 1`)
+)
+
+// SCRAMPassword contains the building blocks to build a PostgreSQL SCRAM
+// verifier. Implements the PostgresPassword interface
+type SCRAMPassword struct {
+	// Hash is the hash algorithm to use. Currently the only thing supported in
+	// PostgreSQL is SHA-256
+	Hash func() hash.Hash
+	// Iterations is the number of iterations to run the PBKDF2 algorithm when
+	// generating the hashed salted password. This defaults to 4096, which is the
+	// PostgreSQL default
+	Iterations int
+	// SaltLength is the length of the generated salt that is used as part of the
+	// PBKDF2 algorithm
+	SaltLength int
+	// password is the plaintext password. This is really the most important
+	// attribute
+	password string
+	// saltGenerator is a function that is used to generate a salt. This can be
+	// mocked for testing purposes
+	saltGenerator saltGenerator
```

Comment:
It's probably simpler to use `saltGenerator func(length int) ([]byte, error)` here.

The mocks can be inline or variables:
```go
happy := func (int) ([]byte, error) {
  return []byte("h1pp0p4rty2020"), nil
}

sad := func (int) ([]byte, error) {
  return nil, ErrSCRAMSaltLengthInvalid
}
```

The constructor can inline the current crypto implementation:
```go
s.saltGenerator = func(length int) ([]byte, error) {
	if length < 1 {
		return []byte{}, ErrSCRAMSaltLengthInvalid
	}

	// create a salt of size length. The slice needs to be allocated as the
	// crypto random number generator copies the byte values into it
	salt := make([]byte, length)
	_, err := rand.Read(salt)
	return salt, err
}
```

Reply:
I was trying to do something like this, but I kept getting panics, likely because I was setting it up incorrectly.

Reply:
Alright got it correct now. I may have been messing up the initialization.